Show Files and Folders in Asset Browser

Sanakan8472 opened the pull request:
![image](https://github.com/ezEngine/ezEngine/assets/11784812/920743ac-8a99-46d8-a8d3-60ed275b98f4)

# Asset Browser Folder View
* Added ability to create, delete, rename and move via drag and drop folders. Press F2 to rename, del to delete a folder.

# Asset Browser
* Any files and folders are now shown in the asset browser view.
* Files can be renamed by pressing F2 and deleted by pressing the del key.
* Assets can be drag & dropped into the folder view to move them around.
* Creating assets via context menu now adds them directly to the view and enters name edit mode. After a name was set, the newly created asset is opened.
* Changed rendering of items:
  * Assets are the only items using alternate background color to easily distinguish them from normal files.
  * Files that can be imported (e.g. tga) have a different icon than regular files (e.g. txt). Files can can be imported but haven't yet turn into an orange icon.

# Asset Import
* Added `ezAssetDocumentGenerator::GetSupportsFileTypes` to create a list of all files that can potentially be imported.
* Asset browser context menu now has an import action. Allowing the import dialog to be opened for the current multi-selection of files.
* Asset browser context menu also shows an *importd via* menu, showing all assets that imported one of the selected files.

# Code Changes
* Created `eqQtAssetBrowserFolderView`, extracted out of `ezQtAssetBrowserWidget`.
* ezQtAssetBrowserFilter
  * `SetShowFilesAndFolders` to toggle the feature on and off.
  * `SetTemporaryPinnedItem` to force an item to be always visible. Used when creating new assets to make sure that they are always visible regardless of type filter etc.
* Drag and drop code simplified byadding QDataStream operators for `T*`, `ezDynamicArray<T>`.
* Added `ezAssetCurator::FindAllUses` and `ezAssetCurator::IsReferenced` for files to quickly check which files are already imported.
* `ezQtAssetBrowserModel::VisibleEntry` added that contains an `ezDataDirPath` which simplifies many of the models data queries.
* Added `ezAssetBrowserItemFlags` to categorize each asset browser item.
* Fixed `ezPathUtils::IsSubPath` and added `ezPathUtils::IsSubPath_NoCase`.
  * Fixed a bug in `ezFileSystemModel` when renaming an item to the same name with additional characters at the end, causing the item to be removed from the model.

# Known issues
* Moving files between data directories does not correctly update the `ezAssetTable`.

Sanakan8472 committed:
Show Files and Folders in Asset Browser

# Asset Browser Folder View
* Added ability to create, delete, rename and move via drag and drop folders. Press F2 to rename, del to delete a folder.

# Asset Browser
* Any files and folders are now shown in the asset browser view.
* Files can be renamed by pressing F2 and deleted by pressing the del key.
* Assets can be drag & dropped into the folder view to move them around.
* Creating assets via context menu now adds them directly to the view and enters name edit mode. After a name was set, the newly created asset is opened.
* Changed rendering of items:
  * Assets are the only items using alternate background color to easily distinguish them from normal files.
  * Files that can be imported (e.g. tga) have a different icon than regular files (e.g. txt). Files can can be imported but haven't yet turn into an orange icon.

# Asset Import
* Added `ezAssetDocumentGenerator::GetSupportsFileTypes` to create a list of all files that can potentially be imported.
* Asset browser context menu now has an import action. Allowing the import dialog to be opened for the current multi-selection of files.
* Asset browser context menu also shows an *importd via* menu, showing all assets that imported one of the selected files.

# Code Changes
* Created `eqQtAssetBrowserFolderView`, extracted out of `ezQtAssetBrowserWidget`.
* ezQtAssetBrowserFilter
  * `SetShowFilesAndFolders` to toggle the feature on and off.
  * `SetTemporaryPinnedItem` to force an item to be always visible. Used when creating new assets to make sure that they are always visible regardless of type filter etc.
* Drag and drop code simplified byadding QDataStream operators for `T*`, `ezDynamicArray<T>`.
* Added `ezAssetCurator::FindAllUses` and `ezAssetCurator::IsReferenced` for files to quickly check which files are already imported.
* `ezQtAssetBrowserModel::VisibleEntry` added that contains an `ezDataDirPath` which simplifies many of the models data queries.
* Added `ezAssetBrowserItemFlags` to categorize each asset browser item.
* Fixed `ezPathUtils::IsSubPath` and added `ezPathUtils::IsSubPath_NoCase`.
  * Fixed a bug in `ezFileSystemModel` when renaming an item to the same name with additional characters at the end, causing the item to be removed from the model.

# Known issues
* Moving files between data directories does not correctly update the `ezAssetTable`.

TheophilusE commented:
Does this account for not deleting things in unmodifiable data directories?

Sanakan8472 commented:
> Does this account for not deleting things in unmodifiable data directories?

I'm not sure what you mean with an unmodified data directory? The change reflects the file system inside the asset browser. If you delete something in the asset browser, it's the same as if you selected the same files in the explorer and deleted them there. Well not quite the same, the editor does a force delete while the explorer moves the files to recycle bin. Maybe I should have a look at that.

TheophilusE commented:
> > Does this account for not deleting things in unmodifiable data directories?
> 
> I'm not sure what you mean with an unmodified data directory? The change reflects the file system inside the asset browser. If you delete something in the asset browser, it's the same as if you selected the same files in the explorer and deleted them there. Well not quite the same, the editor does a force delete while the explorer moves the files to recycle bin. Maybe I should have a look at that.

Yes, but in particular disallow the deletion of assets in non writable special data directories (eg. Base, which is an engine dependency).

jankrassnigg commented:
I don't see that we generally prevent modifications to Base at the moment. We enforce that the data directory is added to the project, but we don't prevent you from changing their content. Another engine that we worked on many years ago did that, and it was quite annoying.

TheophilusE commented:
Oh yeah, that could be annoying. It was just a thought

Sanakan8472 committed:
Review fixes

Fixed bug in new drag and drop code that caused test crash
Use Qfile::moveToTrash instead of deleting files / folder permanently.
Add delete to context menu.
Context menu open now works the same as double clicking an item and has better naming / icons depending on selection content.

jankrassnigg commented:
For follow up work (not necessary in this PR), I'd have a few ideas for more filtering:
* I think we can always show the folders on the right, I would usually just want to hide the files

* I'd like to only see the files that are known to be importable, and hide the rest. This should be an option, something like:
  * Assets Only
  * Assets and Importable Files
  * All Files

* Importable files are already color coded (orange) when they are not imported. Maybe also color code them (green or blue) when they ARE imported. This, together with the icon, makes it easier to distinguish from non-importable files.

* For importable files the "Import" action currently opens the import dialog. For #626 and #429 it would be nice to show a sub-menu with the possible import actions, such as "Import as Mesh" and directly execute that one action for all selected files. This is probably a bit more work to do right.

jankrassnigg commented:
```
--------------------------------------------------------------------
Memory Leak Report:
--------------------------------------------------------------------

Leaked 48 bytes allocated by 'DefaultHeap'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\MemoryTracker.cpp(200):'ezMemoryTracker::AddAllocation'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\Allocator_inl.h(89):'ezInternal::ezAllocatorImpl<ezMemoryPolicies::ezHeapAllocation,7>::Allocate'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(162):'ezDynamicArrayBase<char>::SetCapacity'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(199):'ezDynamicArrayBase<char>::Reserve'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\ArrayBase_inl.h(157):'ezArrayBase<char,ezDynamicArrayBase<char> >::SetCountUninitialized<void>'
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(131):'ezHybridStringBase<32>::operator='
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(42):'ezHybridStringBase<32>::ezHybridStringBase<32>'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserFolderView.cpp(14):'ezFileNameValidator::ezFileNameValidator'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserView.cpp(158):'ezQtIconViewDelegate::createEditor'
(null)(0):'QLineEdit::`default constructor closure''
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserWidget.cpp(969):'ezQtAssetBrowserWidget::NewAsset'
C:\GitHub\ezEngine\Workspace\vs2022x64\Qt6-6.4.0-vs143-x64\include\QtCore\qobjectdefs_impl.h(383):'QtPrivate::QSlotObject<void (__cdecl ezQtAssetBrowserWidget::*)(void),QtPrivate::List<>,void>::impl'
(null)(0):'QFileInfo::filePath'
(null)(0):'QLineEdit::`default constructor closure''
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserWidget.cpp(750):'ezQtAssetBrowserWidget::on_ListAssets_customContextMenuRequested'
C:\GitHub\ezEngine\Workspace\vs2022x64\Code\Editor\EditorFramework\Assets\moc_AssetBrowserWidget.moc.cpp(471):'ezQtAssetBrowserWidget::qt_static_metacall'
C:\GitHub\ezEngine\Workspace\vs2022x64\Code\Editor\EditorFramework\Assets\moc_AssetBrowserWidget.moc.cpp(572):'ezQtAssetBrowserWidget::qt_metacall'
(null)(0):'QFileInfo::filePath'
--------------------------------------------------------------------

Leaked 48 bytes allocated by 'DefaultHeap'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\MemoryTracker.cpp(200):'ezMemoryTracker::AddAllocation'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\Allocator_inl.h(89):'ezInternal::ezAllocatorImpl<ezMemoryPolicies::ezHeapAllocation,7>::Allocate'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(162):'ezDynamicArrayBase<char>::SetCapacity'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(199):'ezDynamicArrayBase<char>::Reserve'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\ArrayBase_inl.h(157):'ezArrayBase<char,ezDynamicArrayBase<char> >::SetCountUninitialized<void>'
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(131):'ezHybridStringBase<32>::operator='
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(42):'ezHybridStringBase<32>::ezHybridStringBase<32>'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserFolderView.cpp(13):'ezFileNameValidator::ezFileNameValidator'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserView.cpp(158):'ezQtIconViewDelegate::createEditor'
(null)(0):'QLineEdit::`default constructor closure''
(null)(0):'QTextCharFormat::setFontLetterSpacingType'
(null)(0):'QLineEdit::`default constructor closure''
C:\GitHub\ezEngine\Code\Editor\EditorFramework\EditorApp\EditorApp.cpp(40):'ezQtEditorApp::RunEditor'
C:\GitHub\ezEngine\Code\Editor\Editor\Editor.cpp(58):'ezEditorApplication::Run'
C:\GitHub\ezEngine\Code\Engine\Foundation\Application\Implementation\MainLoop.cpp(27):'ezRun_MainLoop'
C:\GitHub\ezEngine\Code\Engine\Foundation\Application\Implementation\MainLoop.cpp(67):'ezRun'
C:\GitHub\ezEngine\Code\Engine\Foundation\Application\Implementation\Win\ApplicationEntryPoint_win.h(80):'ezApplicationDetails::ApplicationEntry<ezEditorApplication>'
C:\GitHub\ezEngine\Code\Editor\Editor\Editor.cpp(70):'WinMain'
D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(107):'invoke_main'
D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(288):'__scrt_common_main_seh'
D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(331):'__scrt_common_main'
--------------------------------------------------------------------

Leaked 64 bytes allocated by 'DefaultHeap'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\MemoryTracker.cpp(200):'ezMemoryTracker::AddAllocation'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\Allocator_inl.h(89):'ezInternal::ezAllocatorImpl<ezMemoryPolicies::ezHeapAllocation,7>::Allocate'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(162):'ezDynamicArrayBase<char>::SetCapacity'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(199):'ezDynamicArrayBase<char>::Reserve'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\ArrayBase_inl.h(157):'ezArrayBase<char,ezDynamicArrayBase<char> >::SetCountUninitialized<void>'
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(131):'ezHybridStringBase<32>::operator='
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(42):'ezHybridStringBase<32>::ezHybridStringBase<32>'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserFolderView.cpp(13):'ezFileNameValidator::ezFileNameValidator'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserView.cpp(158):'ezQtIconViewDelegate::createEditor'
(null)(0):'QLineEdit::`default constructor closure''
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserWidget.cpp(969):'ezQtAssetBrowserWidget::NewAsset'
C:\GitHub\ezEngine\Workspace\vs2022x64\Qt6-6.4.0-vs143-x64\include\QtCore\qobjectdefs_impl.h(383):'QtPrivate::QSlotObject<void (__cdecl ezQtAssetBrowserWidget::*)(void),QtPrivate::List<>,void>::impl'
(null)(0):'QFileInfo::filePath'
(null)(0):'QLineEdit::`default constructor closure''
(null)(0):'QLineEdit::`default constructor closure''
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserWidget.cpp(750):'ezQtAssetBrowserWidget::on_ListAssets_customContextMenuRequested'
C:\GitHub\ezEngine\Workspace\vs2022x64\Code\Editor\EditorFramework\Assets\moc_AssetBrowserWidget.moc.cpp(471):'ezQtAssetBrowserWidget::qt_static_metacall'
C:\GitHub\ezEngine\Workspace\vs2022x64\Code\Editor\EditorFramework\Assets\moc_AssetBrowserWidget.moc.cpp(572):'ezQtAssetBrowserWidget::qt_metacall'
(null)(0):'QFileInfo::filePath'
--------------------------------------------------------------------

Leaked 80 bytes allocated by 'DefaultHeap'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\MemoryTracker.cpp(200):'ezMemoryTracker::AddAllocation'
C:\GitHub\ezEngine\Code\Engine\Foundation\Memory\Implementation\Allocator_inl.h(89):'ezInternal::ezAllocatorImpl<ezMemoryPolicies::ezHeapAllocation,7>::Allocate'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(162):'ezDynamicArrayBase<char>::SetCapacity'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\DynamicArray_inl.h(199):'ezDynamicArrayBase<char>::Reserve'
C:\GitHub\ezEngine\Code\Engine\Foundation\Containers\Implementation\ArrayBase_inl.h(157):'ezArrayBase<char,ezDynamicArrayBase<char> >::SetCountUninitialized<void>'
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(131):'ezHybridStringBase<32>::operator='
C:\GitHub\ezEngine\Code\Engine\Foundation\Strings\Implementation\String_inl.h(42):'ezHybridStringBase<32>::ezHybridStringBase<32>'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserFolderView.cpp(13):'ezFileNameValidator::ezFileNameValidator'
C:\GitHub\ezEngine\Code\Editor\EditorFramework\Assets\Implementation\AssetBrowserView.cpp(158):'ezQtIconViewDelegate::createEditor'
(null)(0):'QLineEdit::`default constructor closure''
(null)(0):'QFileInfo::filePath'
(null)(0):'QTextCharFormat::setFontLetterSpacingType'
(null)(0):'QLineEdit::`default constructor closure''
C:\GitHub\ezEngine\Code\Editor\EditorFramework\EditorApp\EditorApp.cpp(40):'ezQtEditorApp::RunEditor'
C:\GitHub\ezEngine\Code\Editor\Editor\Editor.cpp(58):'ezEditorApplication::Run'
C:\GitHub\ezEngine\Code\Engine\Foundation\Application\Implementation\MainLoop.cpp(27):'ezRun_MainLoop'
C:\GitHub\ezEngine\Code\Engine\Foundation\Application\Implementation\MainLoop.cpp(67):'ezRun'
C:\GitHub\ezEngine\Code\Engine\Foundation\Application\Implementation\Win\ApplicationEntryPoint_win.h(80):'ezApplicationDetails::ApplicationEntry<ezEditorApplication>'
C:\GitHub\ezEngine\Code\Editor\Editor\Editor.cpp(70):'WinMain'
D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(107):'invoke_main'
D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(288):'__scrt_common_main_seh'
D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(331):'__scrt_common_main'
--------------------------------------------------------------------

```

Sanakan8472 committed:
Drag and drop into file browser widget. Always show folders. Separate toggle for non-importable files.

Sanakan8472 commented:
Some final fixes:
* Memory leak fixed.
* Folders are always shown.
* Separate toggle for non-importable files in the filter.
* Drag&drop into file browser widget now works.

The rest probably best to make a separate PR for.
